Introduce windows build

# Overview

We tried and some tests are failing and @danielfireman currently does not have a good windows setup to debug/fix them.

I will be happy to help if someone is interested on take this.

---

Please preserve this line to notify @danielfireman (lead of this repository)

I don't know if my setup is "ideal", but I have the latest stable Go, Git and VS Code installed with a workspace for the project. I have a couple of tests failing, for reasons not immediately clear:

(1) In **ExampleLoad_readAll**
```
got:
[]
want:
[[foo] [bar]]
```

(2) In **ExampleLoad_readRaw**
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0xdc1180]

goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc0005cbc68, {0x0, 0x0}, 0x0?, 0x0, {0xe0d600, 0x10fb030})
        C:/Program Files/Go/src/testing/example.go:91 +0x4e5
testing.runExample.func2()
        C:/Program Files/Go/src/testing/run_example.go:59 +0x11c
panic({0xe0d600, 0x10fb030})
        C:/Program Files/Go/src/runtime/panic.go:838 +0x207
github.com/frictionlessdata/datapackage-go/datapackage.ExampleLoad_readRaw()
        d:/Localdev/frictionless/datapackage-go/datapackage/package_test.go:81 +0x300
testing.runExample({{0xe6921a, 0x13}, 0xeaec68, {0xe782f6, 0x2f}, 0x0})
        C:/Program Files/Go/src/testing/run_example.go:63 +0x28d
testing.runExamples(0xc0005cbe58, {0x1102760?, 0x5, 0x1a?})
        C:/Program Files/Go/src/testing/example.go:44 +0x186
testing.(*M).Run(0xc00005b180)
        C:/Program Files/Go/src/testing/testing.go:1721 +0x689
main.main()
        _testmain.go:107 +0x1aa
FAIL    github.com/frictionlessdata/datapackage-go/datapackage  2.663s
```

(3) In **TestLoad**
```
=== RUN   TestLoad/LocalZipWithSubdirs
        d:\Localdev\frictionless\datapackage-go\datapackage\package_test.go:518: [data
                                                                                      oo.csv] != [data/foo.csv]
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -13,7 +13,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: ["ubuntu-latest", "macOS-latest"]
+        os: ["windows-latest","ubuntu-latest", "macOS-latest"]
         go: ["1.15.x", "1.16.x", "1.17.x", "1.18.x"]
     runs-on: ${{ matrix.os }}
 

--- a/datapackage/package.go
+++ b/datapackage/package.go
@@ -10,9 +10,7 @@ import (
 	"io"
 	"io/ioutil"
 	"net/http"
-	"net/url"
 	"os"
-	"path"
 	"path/filepath"
 	"reflect"
 	"strings"
@@ -315,21 +313,6 @@ func Load(path string, loaders ...validator.RegistryLoader) (*Package, error) {
 		return Load(filepath.Join(dir, descriptorFileNameWithinZip), loaders...)
 	}
 	return nil, fmt.Errorf("zip file %s does not contain a file called %s", path, descriptorFileNameWithinZip)
-}
-
-func getBasepath(p string) string {
-	// If it is a remote-like URL, should not treat slashs in a system OS-dependent way.
-	u, err := url.Parse(p)
-	if err == nil && u.IsAbs() {
-		uStr := strings.TrimSuffix(u.String(), "/")
-		uPath := strings.TrimSuffix(u.Path, "/")
-		if uPath == "" {
-			return fmt.Sprintf("%s/", uStr)
-		}
-		return strings.TrimSuffix(uStr, path.Base(u.String()))
-	}
-	// It local path.
-	return filepath.Dir(p)
 }
 
 func read(path string) ([]byte, error) {

--- a/datapackage/package_test.go
+++ b/datapackage/package_test.go
@@ -490,21 +490,23 @@ func TestLoad(t *testing.T) {
 		f, err := w.Create("datapackage.json")
 		is.NoErr(err)
 
+		osPath := filepath.Join("data", "foo.csv")
+		is.NoErr(err)
 		content := fmt.Sprintf(`{
 			"profile": "data-package",
 			"resources": [
 			  {
 				"encoding": "utf-8",
 				"name": "res1",
-				"path": "data%sfoo.csv",
+				"path": "%s",
 				"profile": "data-resource"
 			  }
 			]
-		  }`, string(os.PathSeparator))
+		  }`, osPath)
 		_, err = f.Write([]byte(content))
 		is.NoErr(err)
 		// Writing a file which is in a subdir.
-		f1, err := w.Create("data/foo.csv")
+		f1, err := w.Create(osPath)
 		is.NoErr(err)
 		_, err = f1.Write([]byte(`foo`))
 		is.NoErr(err)
@@ -515,7 +517,7 @@ func TestLoad(t *testing.T) {
 		is.NoErr(err)
 		res := pkg.GetResource("res1")
 		is.Equal(res.name, "res1")
-		is.Equal(res.path, []string{"data/foo.csv"})
+		is.Equal(res.path, []string{osPath})
 		contents, err := res.ReadAll()
 		is.NoErr(err)
 		is.Equal(contents[0], []string{"foo"})

--- a/datapackage/package_test.go
+++ b/datapackage/package_test.go
@@ -491,6 +491,7 @@ func TestLoad(t *testing.T) {
 		is.NoErr(err)
 
 		osPath := filepath.Join("data", "foo.csv")
+		jsonPath, _ := json.Marshal(osPath)
 		is.NoErr(err)
 		content := fmt.Sprintf(`{
 			"profile": "data-package",
@@ -498,11 +499,11 @@ func TestLoad(t *testing.T) {
 			  {
 				"encoding": "utf-8",
 				"name": "res1",
-				"path": "%s",
+				"path": %s,
 				"profile": "data-resource"
 			  }
 			]
-		  }`, osPath)
+		  }`, jsonPath)
 		_, err = f.Write([]byte(content))
 		is.NoErr(err)
 		// Writing a file which is in a subdir.

--- a/datapackage/path.go
+++ b/datapackage/path.go
@@ -1,0 +1,36 @@
+package datapackage
+
+import (
+	"fmt"
+	"net/url"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+func parseRemotePath(path string) (*url.URL, bool) {
+	u, err := url.Parse(path)
+	return u, err == nil && u.Scheme != "" && u.Host != ""
+}
+
+func joinPaths(basePath, finalPath string) string {
+	if u, isRemote := parseRemotePath(basePath); isRemote {
+		u.Path = path.Join(u.Path, finalPath)
+		return u.String()
+	}
+	return filepath.Join(basePath, finalPath)
+}
+
+func getBasepath(p string) string {
+	// If it is a remote-like URL, should not treat slashs in a system OS-dependent way.
+	if u, isRemote := parseRemotePath(p); isRemote {
+		uStr := strings.TrimSuffix(u.String(), "/")
+		uPath := strings.TrimSuffix(u.Path, "/")
+		if uPath == "" {
+			return fmt.Sprintf("%s/", uStr)
+		}
+		return strings.TrimSuffix(uStr, path.Base(u.String()))
+	}
+	// It local path.
+	return filepath.Dir(p)
+}

--- a/datapackage/resource.go
+++ b/datapackage/resource.go
@@ -10,7 +10,6 @@ import (
 	"net/url"
 	"os"
 	"path"
-	"path/filepath"
 	"strings"
 	"sync"
 	"time"
@@ -272,15 +271,6 @@ func loadContents(basePath string, path []string, f func(string) func() (io.Read
 		}
 	}
 	return newMultiReadCloser(rcs), nil
-}
-
-func joinPaths(basePath, path string) string {
-	u, err := url.Parse(basePath)
-	if err != nil {
-		return filepath.Join(basePath, path)
-	}
-	u.Path = filepath.Join(u.EscapedPath(), path)
-	return u.String()
 }
 
 // ReadAll reads all rows from the table and return it as strings.

